File: problems/0014-longest-common-prefix/solution.py
"""
Problem Number: 14
Problem Name: Longest Common Prefix
Difficulty: Easy
Topics: String
Companies: Amazon, Google, Apple
"""

from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
        Approach: Horizontal Scanning
        Time Complexity: O(S)
        Space Complexity: O(1)
        """
        if not strs:
            return ""
            
        prefix = strs[0]
        
        for s in strs[1:]:
            while not s.startswith(prefix):
                prefix = prefix[:-1]
                if not prefix:
                    return ""
                    
        return prefix

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.longestCommonPrefix(["flower","flow","flight"])) # "fl"
    print(solution.longestCommonPrefix(["dog","racecar","car"]))    # ""

File: problems/0014-longest-common-prefix/solution.js
/**
 * @problemNumber: 14
 * @problemName: Longest Common Prefix
 * @difficulty: Easy
 * @topics: String
 * @companies: Amazon, Google, Apple
 */

/**
 * Approach: Horizontal Scanning
 * Time Complexity: O(S) where S is sum of all chars
 * Space Complexity: O(1)
 *
 * @param {string[]} strs
 * @return {string}
 */
function longestCommonPrefix(strs) {
    if (strs.length === 0) return "";

    let prefix = strs[0];

    for (let i = 1; i < strs.length; i++) {
        while (strs[i].indexOf(prefix) !== 0) {
            prefix = prefix.substring(0, prefix.length - 1);
            if (prefix === "") return "";
        }
    }

    return prefix;
}

// Test cases
console.log(longestCommonPrefix(["flower", "flow", "flight"])); // "fl"
console.log(longestCommonPrefix(["dog", "racecar", "car"]));    // ""

module.exports = { longestCommonPrefix };
